acceptance/tests/provider/package/dpkg_ensure_latest_virtual_packages.rb

```diff
@@ -0,0 +1,33 @@
+require 'pry-byebug'
+test_name "dpkg ensure latest with allow_virtual set to true, the virtual package should detect and install a real package"
+confine :to, :platform => /debian/
+tag 'audit:low'
+
+require 'puppet/acceptance/common_utils'
+extend Puppet::Acceptance::PackageUtils
+extend Puppet::Acceptance::ManifestUtils
+pkg = "rubygems"
+step "Ensure latest should install ruby instead of rubygems when allow_virtual is set to true" do
+
+  package_manifest_with_allow_virtual = resource_manifest('package', pkg, ensure: 'latest', allow_virtual: true)
+  apply_manifest_on(agent, package_manifest_with_allow_virtual)
+  output = on(agent.name, "dpkg-query -W --showformat='${Status} ${Package} ${Version} [${Provides}] #{pkg}#' ").output
+  lines = output.split("#")
+  matched_line = lines.find { |package| package.match(/[\[ ](#{Regexp.escape(pkg)})[\],]/)}
+  package_line_info = matched_line.split(" ")
+  real_package_name = package_line_info[3]
+  real_package_installed_version = package_line_info[4]
+  installed_version = on(agent.name, "apt-cache policy #{real_package_name} | sed -n -e 's/Installed: //p'").stdout
+  installed_version.delete!(' ')
+
+  assert_match(real_package_installed_version, installed_version)
+end
+
+step "Ensure latest should not install ruby package if it's allready installed and exit code should be 0" do
+
+  package_manifest_with_allow_virtual = resource_manifest('package', pkg, ensure: 'latest', allow_virtual: true)
+  apply_manifest_on(agent, package_manifest_with_allow_virtual, :detailed_exitcodes => true)
+  exit_code = on(agent.name, "echo $?").output.chop
+
+  assert_match(exit_code, "0")
```

Comment:
Beaker already gives you the exit code of an executed command:
`#<Beaker::Result:0x0000562c2e500e00 @cmd="echo $?", @exit_code=0, @host="ujiay8oe3wy7vzv.delivery.puppetlabs.net", @output="0\n", @raw_output="0\n", @raw_stderr="", @raw_stdout="0\n", @stderr="", @stdout="0\n">`

 You can rewrite this as:
```ruby
apply_manifest_on(agent, resource_manifest('package', pkg, ensure: 'latest', allow_virtual: true), :detailed_exitcodes => true) do
  assert_match(exit_code, 0)
end
```